### ISSUE TITLE
Document rpc node requirements

### ISSUE THREAD

**Issue by rflechtner** [polkadot-api/polkadot-api#565]
As of version 1.13.0 of the kilt-node implementation, I'm not able to connect to a KILT rpc node via WS using polkadot-api.
The error I get is
```
RpcError: invalid type: sequence, expected a (both 0x-prefixed or not) hex string or byte array containing 32 bytes at line 1 column 0
```
so presumably we're dealing with an exception on the side of the node implementation (seems to be happening when calling the method `chainHead_v1_unpin`).

Connections using smoldot seem to be working for the most part, so I suspect this is due to issues in or incompatibilities with the rpc node implementation.

Are there known requirements on minimum polkadot-sdk version ranges etc.? If so, it seems like a good idea to mention that in the docs.

## Steps to reproduce

Try running the following script, you should get an error raised from within `createClient` after 30-60s.
My node version: `v20.13.1`

```ts
import { createClient } from "polkadot-api";
import { WebSocketProvider } from "polkadot-api/ws-provider/node";

// Connect to kilt spiritnet.
const client = createClient(WebSocketProvider("wss://kilt-rpc.dwellir.com"));

// after subscribing and waiting for ~30s, I'm getting the error
client.finalizedBlock$.subscribe((finalizedBlock) =>
  console.log(finalizedBlock.number, finalizedBlock.hash)
);
```

**Comment by josepot** [comment]
> The error I get is
> 
> ```
> RpcError: invalid type: sequence, expected a (both 0x-prefixed or not) hex string or byte array containing 32 bytes at line 1 column 0
> ```
> 
> so presumably we're dealing with an exception on the side of the node implementation (seems to be happening when calling the method `chainHead_v1_unpin`).

This is because your node implements a pretty old version of the -now stable- [`chainSpec` group of functions](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead.html) of the JSON-RPC spec.

> Are there known requirements on minimum polkadot-sdk version ranges etc.? If so, it seems like a good idea to mention that in the docs.

Really good point. We will as soon as we find out.

@jsdw @lexnv @ niklasad1 do you happen to know which version of the polkadot-sdk node is the one that came after [this fix](https://github.com/paritytech/polkadot-sdk/pull/3667) (I think that release is the one that we should use as the minimum polkadot-sdk version node).

I'm guessing that it must be the release of [Polkadot v1.11.0](https://github.com/paritytech/polkadot-sdk/releases/tag/polkadot-v1.11.0), but I'm not certain.

**Comment by rflechtner** [comment]
I was assuming something like that, yeah. 

Spiritnet should be on 1.7.0 soon, it will be a little while until we're updated to 1.11.0 though. We'll try to work with the light client until then. Unfortunately the omni-node is not yet ready to be used by us, because it currently does not support u64 block numbers, which we use on kilt.

If you find ways to support older rpc specs as well, that would make things easier for us of course.

**Comment by carlosala** [comment]
After checking the node implementation a bit, I'm quite certain that a node built with Polkadot equal or after v1.8.0 shouldn't have any issues. [This is the relevant PR.](https://github.com/paritytech/polkadot-sdk/pull/3230) Before that many issues happened, but after v1.8.0 everything should be fine.

**Comment by josepot** [comment]
> After checking the node implementation a bit, I'm quite certain that a node built with Polkadot equal or after v1.8.0 shouldn't have any issues. [This is the relevant PR.](https://github.com/paritytech/polkadot-sdk/pull/3230) Before that many issues happened, but after v1.8.0 everything should be fine.

There were other important bugs that were fixed afterwords, like the one that I linked in my comment... What makes you think that v1.8.0 is the correct target?

**Comment by josepot** [comment]
> If you find ways to support older rpc specs as well, that would make things easier for us of course.

There is a way, actually, which is to create a `JsonRpcProvider` enhancer that handles the upstream issues. I can definitely hook you up. I will try to find some time this week to give you an enhancer that solves these issues.

**Comment by josepot** [comment]
I have been thinking a lot about this and I now realize that we can't release v1 until we don't have a proper solution to this issue.

Therefore, this is what we are going to do do: we will add a new package (possibly named which will provide different enhancers for different versions of the polkadot-sdk node which either have bugs or are not yet using the stable version of the JSON-RPC spec.

What this means is that if you are using a node build with Polkadot equal or after v1.11.0 then you would do:

```ts
import { createClient } from "polkadot-api";
import { WebSocketProvider } from "polkadot-api/ws-provider/node";

const client = createClient(WebSocketProvider("wss://your-fancy-server.io"));
```

However, if your node was built with a version prior to that, then you would have to do something like:

```ts
import { createClient } from "polkadot-api";
import { WebSocketProvider } from "polkadot-api/ws-provider/node";
import compatEnahncer from "polkadot-api/polkadot-sdk-compat";

const client = createClient(
  compatEnhancer(
    WebSocketProvider("wss://your-fancy-server.io")
  )
);
```

The final API may differ a bit (perhaps I will create different enhancers for different versions, but I don't think so). Also, the `compatEnhancer` would have a tiny bit of overhead, but I'm pretty sure that it would be neglectible.

This is not going to be an easy task, but I think that it's pretty important: so I will try to priorize this during this week.

AFWIW @mutantcornholio already requested something like this [a while back](https://github.com/polkadot-api/polkadot-api/issues/515#issuecomment-2112910019). Anyhow, I will get to it right away.

Also, the CLI should use this compat enhancer by default.

**Comment by josepot** [comment]
List of all the enhancers that should be needed:
- remove dups from pruned blocks
- deal with subscription messages out of order
- `finalizedBlockHash` -> `finalizedBlockHashes`
- unpin hashes -> unpin hash
- translate `unstable` methods to `v1` (and the other way around) 
- when [`transaction_v1_broadcast`](https://paritytech.github.io/json-rpc-interface-spec/api/transaction_v1_broadcast.html#transaction_v1_broadcast) is not present use `transaction_unstable_submitAndWatch` as broadcast (same for [`transaction_v1_stop`](https://paritytech.github.io/json-rpc-interface-spec/api/transaction_v1_stop.html#transaction_v1_stop) -> `transactionWatch_unstable_unwatch`)

**Comment by rflechtner** [comment]
Frankly I think that is a very good idea, as polkadot-sdk upgrades have been quite labour-intensive recently, so I would assume we're not the only team that is not on the latest versions - especially given that there's no immediate pressure to upgrade since v1 afaik.

**Comment by josepot** [comment]
> ```ts
> import compatEnhancer from "polkadot-api/polkadot-sdk-compat";
> ```

I finished implementing all these ones, but it turns out that the kilt node is running a super-old version that still uses the old events 😭 E.g:

```json
{
  "jsonrpc":"2.0",
  "method":"chainHead_unstable_call",
  "params":{"subscription":"gDV32FvQCQ7YqfB4","result":{"event":"done","result":"0x080e0000000f000000"}}
}
```
I will try crafting another enhancer to fix this, but I'm afraid that this node version may be too old, so I'm not sure if I will be able to fix all the bugs of this old version just with these enhancers... Let's see.

**Comment by rflechtner** [comment]
> I will try crafting another enhancer to fix these events, but I'm afraid that this node version may be too old, so I'm not sure if I will be able to fix all the bugs of this old version just with these enhancers... Let's see.

We are in the process of updating to 1.7.0, so if the current enhancers work with that version that would be completely sufficient. You could try against branch https://github.com/KILTprotocol/kilt-node/tree/ag_polkadot_v1.7.0 - we have a standalone runtime included, which is pretty easy to spin up (especially when using the `--dev` flag). I could also bundle up a docker image for you to try it against.

**Comment by josepot** [comment]
> > I will try crafting another enhancer to fix these events, but I'm afraid that this node version may be too old, so I'm not sure if I will be able to fix all the bugs of this old version just with these enhancers... Let's see.
> 
> We are in the process of updating to 1.7.0, so if the current enhancers work with that version that would be completely sufficient. You could try against branch https://github.com/KILTprotocol/kilt-node/tree/ag_polkadot_v1.7.0 - we have a standalone runtime included, which is pretty easy to spin up (especially when using the `--dev` flag). I could also bundle up a docker image for you to try it against.

I actually managed to make it work with any node created with polkadot-sdk `1.1.0` or higher, which is a lot more than what I expected! So, basically, if your node is between `1.1.0` and bellow `1.11.0`, then you can use the enhancer provided by the new package to make things work. If your work is equal or higher than `1.11.0` then you won't need any enhancers, and if your node is below `1.1.0` then... tough luck! (at least for now).

It's actually possible (but extremely tricky) to make those below `1.1.0` work, but I mean, `1.1.0` was released around September 2023, so I think that it's fair to say that we don't support a version that's older than that.

**Comment by rflechtner** [comment]
That's fantastic news! Let me know when it's released - I found out that our current node version also seems to be a problem for smoldot (transaction submission and runtime calls don't seem to work), which means that this strategy is the only real option for us to try using papi right now.

**Comment by josepot** [comment]
> That's fantastic news! Let me know when it's released - I found out that our current node version also seems to be a problem for smoldot (transaction submission and runtime calls don't seem to work), which means that this strategy is the only real option for us to try using papi right now.

I would be **very** interested to jump into a call and have a look at the issues that you are facing with smoldot, is that possible? My matrix/element user is `@josepot:matrix.org`, you can ping me there, or send me a DM on twitter (`@Joseptec`). Thanks!

**Comment by josepot** [comment]
reopening this, because it was accidentally closed by #571 ... We still have to release the new version and update the docs
